win: multiple signal handlers

wlemkows opened the pull request:
Support for unlimited number of signal handlers.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1267)

<!-- Reviewable:end -->

lplewa commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/test/signal_handle/signal_handle.c, line 2 at r1](https://reviewable.io:443/reviews/pmem/nvml/1267#-KTxRWBocr202Hoc_0_x:-KTxRWBocr202Hoc_0_y:b-8k88q1) ([raw file](https://github.com/pmem/nvml/blob/91838f003fb41517d78d2256b1ba8a2d4ddead19/src/test/signal_handle/signal_handle.c#L2)):*
> ```C
> /*
>  * Copyright 2014-2016, Intel Corporation
> ```

2016?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/test/unittest/ut_signal.c, line 59 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV03GkIyIHI8qulySYs:-KV03GkIyIHI8qulySYt:bn0hk9i) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L59)):*
> ```C
> 
> int
> find_handler_index(int signum)
> ```

Fix/add comments please.

---

*[src/test/unittest/ut_signal.c, line 136 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV07Ky99G07jAvmvQkx:-KV07KyABliN3IQD2BeS:bgsy9bx) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L136)):*
> ```C
> 	for (int i = 0; i < defined_handlers && handler_assigned != 1; i++) {
> 		if (Sa_handler_tab[i].signum == signum) {
> 			Sa_handler_tab[i].signum = signum;
> ```

Unnecessary.

---

*[src/test/unittest/ut_signal.c, line 147 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV07SAqxqBtsUAkFbI7:-KV07SAqxqBtsUAkFbI8:btcge95) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L147)):*
> ```C
> 			REALLOC(Sa_handler_tab,
> 			sizeof(struct signal_handler) * (defined_handlers + 1));
> 		defined_handlers++;
> ```

Move this two line lower and you won't have to do `defined_handlers - 1`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/test/unittest/ut_signal.c, line 52 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV0AcstSB2BEwQPkta8:-KV0AcstSB2BEwQPkta9:b-pbitaq) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L52)):*
> ```C
> 
> static int defined_handlers = 0;
> static int Signum;
> ```

Is this still needed? Seems like this is only set for SIGSEGV.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 8 of 10 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/test/signal_handle/signal_handle.c, line 2 at r1](https://reviewable.io:443/reviews/pmem/nvml/1267#-KTxRWBocr202Hoc_0_x:-KV_KJPucM53YeRMVKKl:bwluit3) ([raw file](https://github.com/pmem/nvml/blob/91838f003fb41517d78d2256b1ba8a2d4ddead19/src/test/signal_handle/signal_handle.c#L2)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> 2016?
</details>
This is new file created in 2016 so it is ok to have just 2016 in the header.

---

*[src/test/signal_handle/signal_handle.c, line 37 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KlRizLS5f5vDQWtC:-KV_KlRizLS5f5vDQWtD:bsduczf) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L37)):*
> ```C
>  *
>  *
>  * operations are: 's', 'a', 'b', 'i', 'v'
> ```

s: ... SIGSEGV ...
a: ... SIGABRT ...

etc

---

*[src/test/signal_handle/signal_handle.c, line 99 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KeMfCAjYImFoq1a9:-KV_KeMfCAjYImFoq1aA:b-y8gujy) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L99)):*
> ```C
> 			if (!ut_sigsetjmp(Jmp)) {
> 				if (!raise(SIGSEGV)) {
> 					UT_OUT("\t SIGSEGV occurience");
> ```

occurence

---

*[src/test/signal_handle/signal_handle.c, line 106 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KhfumZ4YyIvCKZJo:-KV_KhfumZ4YyIvCKZJp:b-dw6n1q) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L106)):*
> ```C
> 			break;
> 
> 		case 'a':
> ```

what's the difference between 'a' and 'b' ? 

---

*[src/test/unittest/ut_signal.c, line 46 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_L8XrGYPaFJAq5THf:-KV_L8XrGYPaFJAq5THg:b-x253ea) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L46)):*
> ```C
> 
> struct signal_handler
> {
> ```

{ in the same line

---

*[src/test/unittest/ut_signal.c, line 78 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_LIf5245aPXp5XcpK:-KV_LIf6mOQX8AphdXJp:b-yfnld9) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L78)):*
> ```C
> 	if (excode == EXCEPTION_ACCESS_VIOLATION) {
> 		int index = find_handler_index(Signum);
> 		if (index == -1)
> ```

please add {} for both 'if' and 'else' blocks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/test/signal_handle/signal_handle.c, line 37 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KlRizLS5f5vDQWtC:-KVeLX0-oVGiLL4nsYXx:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L37)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> s: ... SIGSEGV ...
> a: ... SIGABRT ...
> 
> etc
</details>
Done.

---

*[src/test/signal_handle/signal_handle.c, line 99 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KeMfCAjYImFoq1a9:-KVeL_02EyDSB-iDt26D:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L99)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> occurence
</details>
Done.

---

*[src/test/signal_handle/signal_handle.c, line 106 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KhfumZ4YyIvCKZJo:-KVd1VgY-1j-nUfnpaxd:b-56xv3v) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L106)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> what's the difference between 'a' and 'b' ? 
</details>
It is case with the same signal occurrence one by one.

---

*[src/test/unittest/ut_signal.c, line 46 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_L8XrGYPaFJAq5THf:-KVeLgtcyvMfkVI1C_Y9:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L46)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> { in the same line
</details>
Done.

---

*[src/test/unittest/ut_signal.c, line 52 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV0AcstSB2BEwQPkta8:-KVZhNRtVlVd95qsX1ci:b-laplul) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L52)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary>
> Is this still needed? Seems like this is only set for SIGSEGV.
</details>
Yes, because for SIGSEGV is special exception_handler pass to AddVectoredExceptionHandler as a param, and there is no way to pass additional signum.

---

*[src/test/unittest/ut_signal.c, line 59 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV03GkIyIHI8qulySYs:-KVeLm0yLlHeugmLgWIf:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L59)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary>
> Fix/add comments please.
</details>
Done.

---

*[src/test/unittest/ut_signal.c, line 78 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_LIf5245aPXp5XcpK:-KVeLpQZMciA8fIqlwrl:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L78)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> please add {} for both 'if' and 'else' blocks
</details>
Done.

---

*[src/test/unittest/ut_signal.c, line 136 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV07Ky99G07jAvmvQkx:-KVeLqdv9ta95RilUAlF:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L136)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary>
> Unnecessary.
</details>
Done.

---

*[src/test/unittest/ut_signal.c, line 147 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV07SAqxqBtsUAkFbI7:-KVeLs8XS2vzRcymYpzR:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L147)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary>
> Move this two line lower and you won't have to do `defined_handlers - 1`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/test/unittest/ut_signal.c, line 52 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV0AcstSB2BEwQPkta8:-KVefTR0-yNNDvQVVTYQ:bubachb) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L52)):*
<details><summary><i>Previously, wlemkows (Weronika Lemkowska) wrote…</i></summary>
> Yes, because for SIGSEGV is special exception_handler pass to AddVectoredExceptionHandler as a param, and there is no way to pass additional signum.
</details>
The only time a see an assignment to `Signum` is when `signum == SIGSEGV`, hence I believe you can live without this static variable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/test/unittest/ut_signal.c, line 52 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV0AcstSB2BEwQPkta8:-KVxgROxnqQn8lhrn-gm:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/unittest/ut_signal.c#L52)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary>
> The only time a see an assignment to `Signum` is when `signum == SIGSEGV`, hence I believe you can live without this static variable.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/test/signal_handle/signal_handle.c, line 106 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KhfumZ4YyIvCKZJo:-KWYJ9W6Z0C2AMO4XqtY:b-qr4sb9) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L106)):*
<details><summary><i>Previously, wlemkows (Weronika Lemkowska) wrote…</i></summary>
> It is case with the same signal occurrence one by one.
</details>
So can't you just execute this test app two times with 'a' ? If there is no difference between the 'a' and 'b' cases I think there is no need for duplicating the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 7 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/test/signal_handle/signal_handle.c, line 106 at r2](https://reviewable.io:443/reviews/pmem/nvml/1267#-KV_KhfumZ4YyIvCKZJo:-KWamsilxmLk8eP1xkmh:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/0e73ac3811b3c9762b4d7f70cb26de546e0c03c2/src/test/signal_handle/signal_handle.c#L106)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> So can't you just execute this test app two times with 'a' ? If there is no difference between the 'a' and 'b' cases I think there is no need for duplicating the code.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Reviewed 2 of 10 files at r1, 5 of 7 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/test/unittest/ut_signal.c, line 144 at r4](https://reviewable.io:443/reviews/pmem/nvml/1267#-KWb-qdJ5sKUboRuH1yv:-KWb-qdKITILkbhD_XkL:b-8tkb7m) ([raw file](https://github.com/pmem/nvml/blob/0d19afaed2950a79a12da52b671ecdc282e3a466/src/test/unittest/ut_signal.c#L144)):*
> ```C
> 		Sa_handler_tab = (struct signal_handler *)
> 			REALLOC(Sa_handler_tab,
> 			sizeof(struct signal_handler) * (defined_handlers + 1));
> ```

This is highly inefficient. I would probably go with `2 * defined_handlers + 1` or something like that. That would of course require altering the handler assignment scheme a bit. Having said that, this is only for unit tests which scarcely use more than one signal handler. As such I won't press to have this changed unless someone supports my comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
@plebioda @krzycz can you answer Tomek comment?

wlemkows commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/test/unittest/ut_signal.c, line 144 at r4](https://reviewable.io:443/reviews/pmem/nvml/1267#-KWb-qdJ5sKUboRuH1yv:-KXnvi357bUTs14huYxo:bmywpe5) ([raw file](https://github.com/pmem/nvml/blob/0d19afaed2950a79a12da52b671ecdc282e3a466/src/test/unittest/ut_signal.c#L144)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

This is highly inefficient. I would probably go with `2 * defined_handlers + 1` or something like that. That would of course require altering the handler assignment scheme a bit. Having said that, this is only for unit tests which scarcely use more than one signal handler. As such I won't press to have this changed unless someone supports my comment.
</blockquote></details>

@plebioda @krzycz can you answer Tomek comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

pbalcer commented on src/test/unittest/ut_signal.c:
```diff
@@ -40,12 +40,31 @@
 
-static void (*Sa_handler) (int signum);
-static int Signum;
+struct signal_handler {
```
This is needlessly complicated. And incorrect.

There are 32 signals that we are interested in, which translates to a static array of 32 sigaction structs and a single static lock;

sigaction:
lock(&action_lock);
if (oldact)
  *oldact = actions[signum - 1];
if (act)
  actions[signum - 1] = *act;
unlock(&action_lock);
return 0;

in the handler code:
actions[signum - 1].sa_handler(signum);

done.

To properly implement sigaction you should also take mask into account make this thing per-thread, but I'm not sure if we want to go that far.

tomaszkapela replied:
I believe the lock should not be necessary, adding signal handlers from multiple threads is not a good thing to do. Adding the lock should not affect anything, but we can add it. Preferably with a trylock and an assert, so that we don't ever try to do this while multithreading.

As for **sa_mask**, AFAIK it is defined per registered handler, which is registered per process, so it shouldn't be held per thread.

pbalcer replied:
Yep, I agree with the sa_mask thing, my bad.

But I think sigaction should be thread-safe ;)

wlemkows commented:
Review status: 5 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/test/unittest/ut_signal.c, line 45 at r4](https://reviewable.io:443/reviews/pmem/nvml/1267#-KTOStkhqHx85fBk30xW-r4-45:-KYrgqES1p8iRcWvKZA4:b-3qc4rg) ([raw file](https://github.com/pmem/nvml/blob/0d19afaed2950a79a12da52b671ecdc282e3a466/src/test/unittest/ut_signal.c#L45)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Yep, I agree with the sa_mask thing, my bad.

But I think sigaction should be thread-safe ;)
</blockquote></details>

I don't know why, I initially assume that we can expect much more signals than 32.
Now I do it in simply way as Piotrek said.
This pull request is only future request, and at this point we probably don't have any test that needs multiple handlers,
especially MT tests, so issue with MT can be resolved another time, if it ever will be necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

plebioda commented:
What are we going to do with this PR ? @wlemkows @krzycz @lplewa

wlemkows commented:
As I remember, from my site it is done.

krzycz commented:
Reviewed 3 of 10 files at r1, 1 of 2 files at r2, 1 of 3 files at r4, 4 of 6 files at r5, 1 of 5 files at r6.
Review status: 7 of 11 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/test/unittest/unittest.h, line 133 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyjpZ3fxJ-AZ6mNM6N:-KdyjpZ3fxJ-AZ6mNM6O:b-q13mg7) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/unittest.h#L133)):*
> ```objc
> #define ut_util_lseek lseek
> #else
> static struct signal_handler *Sa_handler_tab;
> ```

Is this used at all?

---

*[src/test/unittest/ut.c, line 736 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyjmJJZYK1BDWnUlwC:-KdyjmJJZYK1BDWnUlwD:b-vgu1sb) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut.c#L736)):*
> ```C
> 
> #ifdef _WIN32
> 	FREE(Sa_handler_tab);
> ```

Where is it allocated?

---

*[src/test/unittest/ut_signal.c, line 48 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyhogFadU92aaou_rQ:-KdyhogGgATywu3huRAg:b-edv24c) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L48)):*
> ```C
>  * actions[] - allows registering more than one signal/exception handler
>  */
> static struct sigaction actions[32];
> ```

Since this code is Windows-specific, I'd rather use **NSIG** instead of 32, as this is the macro defining maximum signal number on Windows.

---

*[src/test/unittest/ut_signal.c, line 48 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-Kdym9GrBRvexPGjWgIK:-Kdym9GrBRvexPGjWgIL:b-5r5fte) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L48)):*
> ```C
>  * actions[] - allows registering more than one signal/exception handler
>  */
> static struct sigaction actions[32];
> ```

A style nit - pls, capitalize the names  of variables with global scope:
I.e. "Sigactions[]", "Sigactions_lock".

---

*[src/test/unittest/ut_signal.c, line 59 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdylvN_V2m3fFC4fys5:-KdylvN_V2m3fFC4fys6:b-hb401g) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L59)):*
> ```C
> 	DWORD excode = ExceptionInfo->ExceptionRecord->ExceptionCode;
> 	if (excode == EXCEPTION_ACCESS_VIOLATION)
> 		actions[SIGSEGV - 1].sa_handler(SIGSEGV);
> ```

Wouldn't it be easier to have an array of size `NSIG + 1` and get rid of `-1` everywhere?

---

*[src/test/unittest/ut_signal.c, line 97 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyiOXjeA_Vgsvp27EF:-KdyiOXjeA_Vgsvp27EG:ba67co1) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L97)):*
> ```C
> 	return retval;
> #else
> 	pthread_mutex_init(&actions_lock, NULL);
> ```

So, the lock will be initialized each time ut_signal is called? It doesn't look good...
What about moving it to *ut_start* ?
Needless to say - would be a good idea to call *pthread_mutex_destroy* at some point (*ut_done*?)...

---

*[src/test/unittest/ut_signal.c, line 99 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdylefLRZG56BtZ9X7n:-KdylefMG-pBvtbCd2Um:b-j76in8) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L99)):*
> ```C
> 	pthread_mutex_init(&actions_lock, NULL);
> 	if (signum > 32)
> 		signum = 32;
> ```

Perhaps it's better to put an assert on `signum < NSIG` instead of handling it that way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 4 of 11 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/test/unittest/unittest.h, line 133 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyjpZ3fxJ-AZ6mNM6N:-Ke3MHplkS9ylU5SyI6Q:bss4a2b) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/unittest.h#L133)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is this used at all?
</blockquote></details>

Removed.

---

*[src/test/unittest/ut.c, line 736 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyjmJJZYK1BDWnUlwC:-Ke3MKYt3lAheIADZvME:bss4a2b) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut.c#L736)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Where is it allocated?
</blockquote></details>

Removed.

---

*[src/test/unittest/ut_signal.c, line 48 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyhogFadU92aaou_rQ:-Ke3MbzZ9qpTX4UUP-ej:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L48)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Since this code is Windows-specific, I'd rather use **NSIG** instead of 32, as this is the macro defining maximum signal number on Windows.
</blockquote></details>

Done.

---

*[src/test/unittest/ut_signal.c, line 48 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-Kdym9GrBRvexPGjWgIK:-Ke3OcivAJda8wecyjbx:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L48)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

A style nit - pls, capitalize the names  of variables with global scope:
I.e. "Sigactions[]", "Sigactions_lock".
</blockquote></details>

Done.

---

*[src/test/unittest/ut_signal.c, line 59 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdylvN_V2m3fFC4fys5:-Ke3X9B6C-iKcdW7iDNd:b7ow95a) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L59)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Wouldn't it be easier to have an array of size `NSIG + 1` and get rid of `-1` everywhere?
</blockquote></details>

Ok done. NSIG will be enough, because NSIG definition is maximum signal number + 1

---

*[src/test/unittest/ut_signal.c, line 97 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdyiOXjeA_Vgsvp27EF:-Ke3yIbs00WSAekxKyxI:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L97)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

So, the lock will be initialized each time ut_signal is called? It doesn't look good...
What about moving it to *ut_start* ?
Needless to say - would be a good idea to call *pthread_mutex_destroy* at some point (*ut_done*?)...

</blockquote></details>

Done.

---

*[src/test/unittest/ut_signal.c, line 99 at r6](https://reviewable.io:443/reviews/pmem/nvml/1267#-KdylefLRZG56BtZ9X7n:-Ke3YuOU0WdhZKDEqohc:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/caf135bccbf945eb11e0f3087bd4a0f02d43ff3b/src/test/unittest/ut_signal.c#L99)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Perhaps it's better to put an assert on `signum < NSIG` instead of handling it that way.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[src/test/unittest/ut_signal.c, line 144 at r4](https://reviewable.io:443/reviews/pmem/nvml/1267#-KWb-qdJ5sKUboRuH1yv:-Ke491VJJUT_v3Sc6mRO:bt1pozb) ([raw file](https://github.com/pmem/nvml/blob/0d19afaed2950a79a12da52b671ecdc282e3a466/src/test/unittest/ut_signal.c#L144)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

@plebioda @krzycz can you answer Tomek comment?
</blockquote></details>

I'm not sure I understand his comment... ;-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267#-:-Ke49IdkcRqJrW1PdfPU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 9 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/test/unittest/ut_signal.c, line 144 at r4](https://reviewable.io:443/reviews/pmem/nvml/1267#-KWb-qdJ5sKUboRuH1yv:-Ke8-UXE1TBRPdc6nXu3:baxai1s) ([raw file](https://github.com/pmem/nvml/blob/0d19afaed2950a79a12da52b671ecdc282e3a466/src/test/unittest/ut_signal.c#L144)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

I'm not sure I understand his comment... ;-)
</blockquote></details>

It is outdated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1267)*
<!-- Sent from Reviewable.io -->
